src/stack.cpp

```diff
@@ -162,115 +162,59 @@ int print_stack_command(int argc, char** argv, Output output) {
     HSTag* tag;
-    if (slice_highest_layer(s) != data->layer) {
+    if (slice_highest_layer(s) != layer) {
         /** slice only is added to its highest layer.
          * just skip it if the slice is not shown on this data->layer */
-        return;
+        return {};
     }
     switch (s->type) {
         case SLICE_CLIENT:
-            if (data->len) {
-                if (data->real_clients) {
-                    data->buf[0] = s->data.client->x11Window();
-                } else {
-                    data->buf[0] = s->data.client->decorationWindow();
-                }
-                data->buf++;
-                data->len--;
+            if (real_clients) {
+                return { s->data.client->x11Window() };
             } else {
-                data->missing++;
+                return { s->data.client->decorationWindow() };
             }
             break;
         case SLICE_WINDOW:
-            if (!data->real_clients) {
-                if (data->len) {
-                    data->buf[0] = s->data.window;
-                    data->buf++;
-                    data->len--;
-                } else {
-                    data->missing++;
-                }
+            if (!real_clients) {
+                return { s->data.window };
             }
             break;
         case SLICE_MONITOR:
             tag = s->data.monitor->tag;
-            if (!data->real_clients) {
-                if (data->len) {
-                    data->buf[0] = s->data.monitor->stacking_window;
-                    data->buf++;
-                    data->len--;
-                } else {
-                    data->missing++;
-                }
-            }
-            int remain_len = 0; /* remaining length */
-            tag->stack->toWindowBuf(data->buf, data->len,
-                                      data->real_clients, &remain_len);
-            int len_used = data->len - remain_len;
-            if (remain_len >= 0) {
-                data->buf += len_used;
-                data->len = remain_len;
-            } else {
-                data->len = 0;
-                data->missing += -remain_len;
+            vector<Window> result;
+            if (!real_clients) {
+                result.push_back(s->data.monitor->stacking_window);
             }
+            vector_append(result, tag->stack->toWindowBuf(real_clients));
+            return result;
             break;
     }
+    return {};
 }
 
-void Stack::toWindowBuf(Window* buf, int len,
-                         bool real_clients, int* remain_len) {
-    struct s2wb data = {};
-    data.len = len;
-    data.buf = buf;
-    data.missing = 0;
-    data.real_clients = real_clients;
-
+vector<Window> Stack::toWindowBuf(bool real_clients) {
+    vector<Window> result;
     for (int i = 0; i < LAYER_COUNT; i++) {
-        data.layer = (HSLayer)i;
         for (auto slice : top[i]) {
-            slice_to_window_buf(slice, &data);
+            vector_append(
```

Comment:
Maybe it's just a matter of personal preference, but I think I would have gone with just another `for` loop (calling `push_back`) instead of introducing a helper like that.

Reply:
I need this `vector_append` at multiple places (4 times). I don't think that it makes the code better if I replace the 4 calls of `vector_append` with a for loop each.